File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> arr = new ArrayList<>();
        int n1 = matrix.length;
        int n2 = matrix[0].length;
        int a = 0;
        int b = 0;
        while(n1 > 0 && n2 > 0){
            for(int i = b; i < n2+b; i++){
                arr.add(matrix[a][i]); 
            }
            for(int i = a+1; i <a+n1; i++){
                arr.add(matrix[i][b+n2-1]);
            }
            if(n1>1){
                for(int i = b+n2-2; i >= b; i--){
                    arr.add(matrix[a+n1-1][i]);
                }
            }
            if(n2>1){
                for(int i = a+n1-2; i > a; i--){
                    arr.add(matrix[i][b]);
                }
            }
            a++;
            b++;
            n1 -= 2;
            n2 -= 2;
        }
        return arr;
    }
}

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        r = len(matrix)
        c = len(matrix[0])

        r0, r1 = 0,r-1
        c0, c1 = 0, c-1
        arr = []
        while(len(arr) < r*c):
            for i in range(c0,c1+1):
                arr.append(matrix[r0][i])    
            r0 += 1
            if(len(arr) == r*c):
                break
            for j in range(r0,r1+1):
                arr.append(matrix[j][c1])
            c1 -= 1
            if(len(arr) == r*c):
                break
            for k in range(c1,c0-1,-1):
                arr.append(matrix[r1][k])
            r1 -= 1
            if(len(arr) == r*c):
                break
            for l in range(r1,r0-1,-1):
                arr.append(matrix[l][c0])
            c0 += 1

        return arr
        
            

